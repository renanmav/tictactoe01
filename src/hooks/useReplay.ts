import { useEffect } from 'react';
import { Alert } from 'react-native';

import { Winner } from './useWinner';

export function useReplay(winner: Winner, restart: () => void) {
  useEffect(() => {
    if (winner !== '?') {
      Alert.alert(
        winner === 'X' ? 'Parab√©ns' : 'Que pena',
        winner === 'C' ? 'Deu velha' : `Jogador ${winner} ganhou`,
        [
          {
            text: 'Jogar novamente',
            onPress: restart,
          },
        ]
      );
    }
  }, [winner, restart]);
}
