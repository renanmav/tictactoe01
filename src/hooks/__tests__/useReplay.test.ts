import { renderHook } from '@testing-library/react-hooks';
import { FetchMock } from 'jest-fetch-mock';

import { useReplay } from '../useReplay';

jest.mock('react-native', () => ({
  Alert: {
    alert: jest.fn(),
  },
}));

import { Alert } from 'react-native';

beforeEach(() => {
  (fetch as FetchMock).resetMocks();
  jest.clearAllMocks();
});

describe('useReplay', () => {
  test('if nothing is shown when there is no winner', () => {
    const restart = jest.fn();

    renderHook(() => useReplay('?', restart));

    expect(restart).not.toBeCalled();
    expect(Alert.alert).not.toBeCalled();
  });

  test('if Alert.alert is called when the winner is "X"', () => {
    const restart = jest.fn();

    renderHook(() => useReplay('X', restart));

    expect(restart).not.toBeCalled();
    expect(Alert.alert).toBeCalledTimes(1);
    expect((Alert.alert as jest.Mock).mock.calls).toMatchSnapshot();
  });

  test('if Alert.alert is called when the winner is "O"', () => {
    const restart = jest.fn();

    renderHook(() => useReplay('O', restart));

    expect(restart).not.toBeCalled();
    expect(Alert.alert).toBeCalledTimes(1);
    expect((Alert.alert as jest.Mock).mock.calls).toMatchSnapshot();
  });

  test('if Alert.alert is called when the winner is "C" (for cat)', () => {
    const restart = jest.fn();

    renderHook(() => useReplay('C', restart));

    expect(restart).not.toBeCalled();
    expect(Alert.alert).toBeCalledTimes(1);
    expect((Alert.alert as jest.Mock).mock.calls).toMatchSnapshot();
  });

  test('if restart is called when Alert.alert is pressed', () => {
    const restart = jest.fn();

    renderHook(() => useReplay('X', restart));

    (Alert.alert as jest.Mock).mock.calls[0][2][0].onPress();

    expect(restart).toBeCalledTimes(1);
  });
});
