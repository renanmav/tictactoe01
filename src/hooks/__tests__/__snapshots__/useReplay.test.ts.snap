// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useReplay if Alert.alert is called when the winner is "C" (for cat) 1`] = `
Array [
  Array [
    "Que pena",
    "Deu velha",
    Array [
      Object {
        "onPress": [MockFunction],
        "text": "Jogar novamente",
      },
    ],
  ],
]
`;

exports[`useReplay if Alert.alert is called when the winner is "O" 1`] = `
Array [
  Array [
    "Que pena",
    "Jogador O ganhou",
    Array [
      Object {
        "onPress": [MockFunction],
        "text": "Jogar novamente",
      },
    ],
  ],
]
`;

exports[`useReplay if Alert.alert is called when the winner is "X" 1`] = `
Array [
  Array [
    "Parab√©ns",
    "Jogador X ganhou",
    Array [
      Object {
        "onPress": [MockFunction],
        "text": "Jogar novamente",
      },
    ],
  ],
]
`;
